HPCC-16946 Add server list handling to CLocalEnvironment

AttilaVamos opened the pull request:
Reorganise checking process

Add Drop Zone server list checking if the target machine doesn't match
to local one.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.13 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1620,11 +1612,57 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+    }
+    if (dropZone)
+    {
+        // check if it is the local machine
```
local to environment?

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1620,11 +1612,57 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+    {
+        // check if it is the local machine
+        Owned<IConstMachineInfo> machineInfo = getMachineByAddress(netaddress);
+        if (machineInfo)
```
I don't think this section is right.. (but I may misunderstand)

As I understand it, dropzones will be defined like this in future:
```
  <DropZone build="_"
            buildSet="DropZone"
            computer=""
            description="DropZone process"
            directory="/var/lib/HPCCSystems/mydropzone"
            ECLWatchVisible="true"
            name="dropzone_2"
            umask="022">
   <WhiteList name="WhiteList" server="2.1.1.1"/>
   <WhiteList name="WhiteList_2" server="2.4.4.4"/>
  </DropZone>
```

i.e. WhiteList - but no defined @computer name (shouldn't be there at all in the definition I'd have thought).

DropZone's with @computer are therefore legacy definitions - which we have to continue to support for backward compatibility reasons only.

So this section should be dealing with the legacy case afaics, i.e. checking if found dropzone has a @computer and if it's netAddress matches.

jakesmith reviewed (changes requested):
@AttilaVamos - please see comments/questions.

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.145 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |


WUtools -selftest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|19 |19 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.39 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |


WUtools -selftest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|19 |19 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.443 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |


WUtools -selftest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|19 |19 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site
| errors | warnings | build time | 
--------------------
| 0 | 65 | 90.268 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:
| Test | total | passed | failed | errors | timeout | 
--------------------
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site
| errors | warnings | build time | 
--------------------
| 0 | 65 | 91.502 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:
| Test | total | passed | failed | errors | timeout | 
--------------------
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site
| errors | warnings | build time | 
--------------------
| 0 | 65 | 91.502 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:
| Test | total | passed | failed | errors | timeout | 
--------------------
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|------|-----------|
| 0 | 65 | 91.502 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:
| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|-----|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|------|-----------|
| 0 | 65 | 91.502 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|-----|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 89.526 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x: :white_check_mark: 
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|------|-----------|
| 0 | 65 | 91.502 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|-----|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site
| errors | warnings | build time | 
--------------------
| 0 | 65 | 90.897 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:
| Test | total | passed | failed | errors | timeout | 
--------------------
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0c3398aa0ce603c5fc53cf7aa80a901be74ccf23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 89.894 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: b5b73b29508ad78df5d610316e47c9b09f48feef
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 137.65 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0xenial_amd64.deb
HPCC Start: Fail
HPCC start:
Exception: 2:
Error - syntax error "Mismatched opening and closing tags" [line 26, file offset 804]
="ServerList" server="localhost"
  </D*ERROR*ropZone>
  <RoxieCluster allFilesDynamic
There are no components configured to run on this node ...
HPCC status:
Exception: 2:
Error - syntax error "Mismatched opening and closing tags" [line 26, file offset 804]
="ServerList" server="localhost"
  </D*ERROR*ropZone>
  <RoxieCluster allFilesDynamic
There are no components configured to run on this node ...

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: b5b73b29508ad78df5d610316e47c9b09f48feef
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 136.31 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0xenial_amd64.deb
HPCC Start: Fail
HPCC start:
Exception: 2:
Error - syntax error Mismatched opening and closing tags [line 26, file offset 804]
=ServerList server=localhost
  </D*ERROR*ropZone>
  <RoxieCluster allFilesDynamic
There are no components configured to run on this node ...
HPCC status:
Exception: 2:
Error - syntax error Mismatched opening and closing tags [line 26, file offset 804]
=ServerList server=localhost
  </D*ERROR*ropZone>
  <RoxieCluster allFilesDynamic
There are no components configured to run on this node ...

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -1580,24 +1580,16 @@ void CLocalEnvironment::clearCache()
 #endif
-
+    // Check the directory path first
     SCMStringBuffer dropzonePath;
```
This comment didn't address in the latest commit.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 364b3f03804d8c1e25a922ecf5aac4974e4b68d4
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.331 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith reworked, please review.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+            dropZone->getComputerName(dropZoneMachineName);
+
+            // check if it is local machine
+            if (strcmp(netaddress, dropZoneMachineName.str()) !=  0)
```
This won't work/doesn't maintain backward compatibility. netaddress can't be compared to legacy @computer name...

Previously, netAddress was used to lookup a IConstMachineInfo with getMachineByAddress, then the name of that computer was matched against a DZ.
You still need to do that to maintain backward compatibility.

So _if_ getComputerName returns anything - then it is an old definition and you should perform same checks as before.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
         const char * pdropzoneDir = fullDropZoneDir.str();
 
-        // Check target file path starts with this Drop zone path
-        // the drop zone paths can be nested (nothing forbids it) like
```
This functionality seems to have been lost?

AttilaVamos replied:
More or less, yes. That was an approach where we have one set of Drop Zones and each assigned only one machine. Now we can have a set of Drop Zones and set of machines for each zones. I think there is no real sense to search a shortest matching drop zone because it should match the server as well. So now if the drop zone directory match to the first part of target filename, then it checks computer (if it is exists - backward compatibility) or the server list. If there is a match then the operation gets green light. If the server doesn't match to any of the server list then the check (outside loop) goes to the next drop zone.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
-                retVal = &zoneIt->query();
+                // It isn't local machine or it has not '@server' attrib. Check the ServerList
+                bool isServerFound = false;
+                Owned<IPropertyTreeIterator> whiteListIt = dropZone->getPTree().getElements("ServerList");
```
Should extend IConstDropZoneInfo instead of accessing IPT directly.
e.g. IDropZoneServerIterator *serverIter = dropZone->getServers();

jakesmith replied:
NB: @wangkx - will also need to use this interface when retrieving the DZ's for esp services.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+                Owned<IPropertyTreeIterator> whiteListIt = dropZone->getPTree().getElements("ServerList");
+                ForEach(*whiteListIt)
+                {
+                    const char *server = whiteListIt->query().queryProp("@server");
```
As previously mentioned, should rename all references to "white list" and use "server list" to avoid confusion, including in JIRA and PR headings

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+#ifdef _DEBUG
+                    LOG(MCdebugInfo, unknownJob, "Listed server: '%s', IP: '%s'", server, serverIpString.str());
+#endif
+                    if (strcmp(netaddress,serverIpString.str()) ==  0)
```
Better to use IpAddress.ipequals() - i.e. make another IpAddress for netaaddress

jakesmith reviewed (changes requested):
@AttilaVamos - added some comments.

Also need to think about adding a switch to disable checking via environment.conf, but that can be separate PR so as not to hold this one up.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -1580,24 +1580,16 @@ void CLocalEnvironment::clearCache()
+    // Check the directory path first
     SCMStringBuffer dropzonePath;
     const char * pdropzonePath = nullptr;
     unsigned dropzonePathLen = _MAX_PATH + 1;
```
Removed

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -1580,24 +1580,16 @@ void CLocalEnvironment::clearCache()
     const char * pdropzonePath = nullptr;
     unsigned dropzonePathLen = _MAX_PATH + 1;
 
+    Owned<IConstDropZoneInfoIterator> zoneIt= getDropZoneIterator();
```
Changed

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+
+            // For backward compatibility
+            SCMStringBuffer dropZoneMachineName;
+            // Returns dropzone '@server' value if it is exists
```
Changed

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d5ee2a05387704c07e06aa9e02b202152c3a3215
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.061 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith updated, please re-review.

jakesmith commented:
@AttilaVamos - haven't looked yet, but you may have missed prev. comment above:

>As previously mentioned, should rename all references to "white list" and use "server list" to avoid confusion, including in JIRA and PR headings

Can you change the PR comment heading and JIRA heading to refer to server list instead of white list.

AttilaVamos commented:
Whitelist changed to server list in JIRA's, and this PR's name.

richardkchapman commented:
@jakesmith I think this is back with you

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -913,6 +932,31 @@ class CConstInstanceInfo : public CConstEnvBase, implements IConstInstanceInfo
+    }
+    virtual StringBuffer & getServer( StringBuffer & server) const
+    {
+        server.set(prop->queryProp("@server"));
```
the convention on StringBuffer &get* methods is to append to the buffer, so that callers can fill up a buffer as they see fit.
i.e.:
```
    virtual StringBuffer & getServer( StringBuffer & server) const
    {
        return server.append(prop->queryProp("@server"));
    }
```

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+                    LOG(MCdebugInfo, unknownJob, "Drop zone coputer: '%s', IP: '%s'", dropZoneMachineName.str(), dropZoneMachineNetAddressIpString.str());
+#endif
+                    if (targetIp.ipequals(dropZoneMachineIp))
+                       break;
```
see comment above, think it should still continue, remembering current match path length, for backward compatibility reasons.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+                    {
+                        // OK the target is a valid machine on the white list
+                        isServerFound = true;
+                        break;
```
see comment above, think it should still continue, remembering current match path length, for backward compatibility reasons.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+            SCMStringBuffer dropZoneMachineName;
+            // Returns dropzone '@computer' value if it is exists
+            dropZone->getComputerName(dropZoneMachineName);
+            if ( 0 != dropZoneMachineName.length())
```
formatting: spurious space before 0

AttilaVamos replied:
Removed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+            }
+
+            // check if it is local machine
+            //if (strcmp(netaddress, dropZoneMachineName.str()) !=  0)
```
```
    // check if it is local machine
    //if (strcmp(netaddress, dropZoneMachineName.str()) !=  0)
```
'local' isn't correct here, these servers can be anywhere.
Maybe these are old lines from previous PR version?

Should be removed I think + newline before else

AttilaVamos replied:
Removed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+#endif
+                    if (targetIp.ipequals(serverIP))
+                    {
+                        // OK the target is a valid machine on the white list
```
... in the server list.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1732,6 +1822,51 @@ unsigned CConstDropZoneIteratorByComputer::count() const
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    serverListIt.setown(dropZone->getPTree().getElements("ServerList"));
```
will leak IPT here, because getPTree() returns a linked tree.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1732,6 +1822,51 @@ unsigned CConstDropZoneIteratorByComputer::count() const
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    serverListIt.setown(dropZone->getPTree().getElements("ServerList"));
+    maxIndex = dropZone->getPTree().getCount("ServerList");
```
leaks IPT

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1732,6 +1822,51 @@ unsigned CConstDropZoneIteratorByComputer::count() const
+}
+
+bool CConstDropZoneServerInfoIterator::isValid()
+{
```
is curr is set (by first() or next()) then it's valid, so would be more correct to do return nullptr != curr, instead of code below.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1732,6 +1822,51 @@ unsigned CConstDropZoneIteratorByComputer::count() const
+{
+    bool hasFirst = serverListIt->first();
+    if (hasFirst)
+        curr = new CConstDropZoneServerInfo(constEnv, &serverListIt->query());
```
need to be careful with existing 'curr' , e.g. if first() recalled.

It should release curr.
Probably easiest if you make curr a Owned and setown here.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1732,6 +1822,51 @@ unsigned CConstDropZoneIteratorByComputer::count() const
+{
+    bool hasNext = serverListIt->next();
+    if (hasNext)
+        curr = new CConstDropZoneServerInfo(constEnv, &serverListIt->query());
```
this will leak existing curr, similar to comment in first() above.

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.hpp:
```diff
@@ -104,13 +104,26 @@ interface IConstInstanceInfo : extends IConstEnvBase
 
+interface IConstDropZoneServerInfo : extends IConstEnvBase
+{
+    virtual StringBuffer & getName( StringBuffer & name) const = 0;
```
formatting: space before StringBuffer

AttilaVamos replied:
Removed.

jakesmith commented on common/environment/environment.hpp:
```diff
@@ -104,13 +104,26 @@ interface IConstInstanceInfo : extends IConstEnvBase
+interface IConstDropZoneServerInfo : extends IConstEnvBase
+{
+    virtual StringBuffer & getName( StringBuffer & name) const = 0;
+    virtual StringBuffer & getServer( StringBuffer & server) const = 0;
```
formatting: space before StringBuffer

AttilaVamos replied:
Removed.

jakesmith reviewed (changes requested):
@AttilaVamos - looking close.

jakesmith commented:
Hm, gitbot still left this with me despite my last posted comment.

@AttilaVamos - looking close.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -59,6 +59,25 @@ class CConstMachineInfoIterator : public  CSimpleInterfaceOf<IConstMachineInfoIt
+protected:
+    IConstDropZoneServerInfo * curr = nullptr;
+    Owned<CLocalEnvironment> constEnv;
+    Owned<IPropertyTreeIterator> serverListIt = nullptr;
```
Removed

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -81,7 +100,7 @@ class CConstDropZoneIteratorByComputer : public CSimpleInterfaceOf<IConstDropZon
 public:
     CConstDropZoneIteratorByComputer(const char * computer);
-
+    ~CConstDropZoneIteratorByComputer() { if (curr != nullptr) delete curr; };
```
Removed.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -913,6 +932,31 @@ class CConstInstanceInfo : public CConstEnvBase, implements IConstInstanceInfo
+    IMPLEMENT_ICONSTENVBASE;
+    CConstDropZoneServerInfo(CLocalEnvironment *env, IPropertyTree *root) : CConstEnvBase(env, root), prop(root) {}
+
+    virtual StringBuffer & getName( StringBuffer & name) const
```
Removed.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -913,6 +932,31 @@ class CConstInstanceInfo : public CConstEnvBase, implements IConstInstanceInfo
+
+    virtual StringBuffer & getName( StringBuffer & name) const
+    {
+        name.set(prop->queryProp("@name"));
```
Changed.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -913,6 +932,31 @@ class CConstInstanceInfo : public CConstEnvBase, implements IConstInstanceInfo
+        name.set(prop->queryProp("@name"));
+        return name;
+    }
+    virtual StringBuffer & getServer( StringBuffer & server) const
```
Removed.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -913,6 +932,31 @@ class CConstInstanceInfo : public CConstEnvBase, implements IConstInstanceInfo
+    }
+
+private:
+    CConstDropZoneServerInfo();
```
Removed.

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -945,6 +989,11 @@ class CConstDropZoneInfo : public CConstEnvBase, implements IConstDropZoneInfo
     }
+    virtual IConstDropZoneServerInfoIterator * getServers() const
+    {
+        // TODO
```
Removed.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d73465886921aa05d200ffddb744196a02643db3
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.286 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |16 |3 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith Updated, comments are addressed/answered, please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: db3cabc6356b963b1b2d7038f03014b48e54bcc7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.54 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |16 |3 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: db3cabc6356b963b1b2d7038f03014b48e54bcc7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 95.712 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:


|
 
p
h
a
s
e
 
|
 
t
o
t
a
l
 
|
 
p
a
s
s
 
|
 
f
a
i
l
 
|
 


|
-
-
-
|
-
-
-
|
-
-
-
|
-
-
-
|


|
s
e
t
u
p
 
(
h
t
h
o
r
)
 
|
9
 
|
9
 
|
0
 
|


|
s
e
t
u
p
 
(
t
h
o
r
)
 
|
9
 
|
9
 
|
0
 
|


|
s
e
t
u
p
 
(
r
o
x
i
e
)
 
|
9
 
|
9
 
|
0
 
|


|
t
e
s
t
 
(
h
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
r
o
x
i
e
)
 
|
0
 
|
0
 
|
0
 
|




HPCC Stop: OK
HPCC Uninstall: OK

 ... ( comment is too long, 240 lines, 1210 bytes, truncated) ... 
ns
t
 
(
h
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
r
o
x
i
e
)
 
|
0
 
|
0
 
|
0
 
|




HPCC Stop: OK
HPCC Uninstall: OK
```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: db3cabc6356b963b1b2d7038f03014b48e54bcc7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.792 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:


|
 
p
h
a
s
e
 
|
 
t
o
t
a
l
 
|
 
p
a
s
s
 
|
 
f
a
i
l
 
|
 


|
-
-
-
|
-
-
-
|
-
-
-
|
-
-
-
|


|
s
e
t
u
p
 
(
h
t
h
o
r
)
 
|
9
 
|
9
 
|
0
 
|


|
s
e
t
u
p
 
(
t
h
o
r
)
 
|
9
 
|
9
 
|
0
 
|


|
s
e
t
u
p
 
(
r
o
x
i
e
)
 
|
9
 
|
9
 
|
0
 
|


|
t
e
s
t
 
(
h
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
r
o
x
i
e
)
 
|
0
 
|
0
 
|
0
 
|




HPCC Stop: OK
HPCC Uninstall: OK

 ... ( comment is too long, 240 lines, 1210 bytes, truncated) ... 
ns
t
 
(
h
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
t
h
o
r
)
 
|
0
 
|
0
 
|
0
 
|


|
t
e
s
t
 
(
r
o
x
i
e
)
 
|
0
 
|
0
 
|
0
 
|




HPCC Stop: OK
HPCC Uninstall: OK
```

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1668,22 +1666,23 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
                     LOG(MCdebugInfo, unknownJob, "Drop zone coputer: '%s', IP: '%s'", dropZoneMachineName.str(), dropZoneMachineNetAddressIpString.str());
```
typo: coputer->computer

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1648,13 +1646,13 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
 
         if (strncmp(pdropzoneDir, targetFilePath, strlen(pdropzoneDir)) == 0)
```
trivial: no need for pdropzonedir var. or strlen, could do strncmp(fullDropZoneDir, targetFilePath, fullDropZoneDir.length())

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1630,7 +1626,9 @@ void CLocalEnvironment::clearCache()
 IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * netaddress, const char *targetFilePath) const
 {
     IConstDropZoneInfo * dropZone = nullptr;
+    IConstDropZoneInfo * candidateDropZone = nullptr;
```
trivial: slightly clearer if declared inside strncmp block, as set at top and not used outside that block.

AttilaVamos replied:
Moved.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1668,22 +1666,23 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
-                }
+                    {
+                        // Keep this drop zone
+                       if (dropzonePathLen > fullDropZoneDir.length())
```
Should this be >= ?

AttilaVamos replied:
I think if the dropzonePathLen is equal to fullDropZoneDir.length() then it means (to me) we already have that path. I don't think it can be add any further information to store same value into the variables in the conditional block. (Perhaps same drop zone name can't appear in different DropZone defininition, but I'm not sure.)
Why do you think it would be useful?

jakesmith replied:
I was thinking about the 1st set, but ignoring fact that the initial value of dropzonePathLen is set at high max.

So yes, it's fine/correct as it is.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1698,16 +1697,18 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+                    if (dropzonePathLen > fullDropZoneDir.length())
+                    {
+                       dropzonePathLen = fullDropZoneDir.length();
+                       dropZone = candidateDropZone;
```
Is there any reason to have this condition code after the ForEach loop above and not do it at the time (instead of) setting isServerFound ?
i.e.
```
    if (targetIp.ipequals(serverIP))
    {
        // Keep this candidate drop zone if its directory path is shorter than we already have
        if (dropzonePathLen > fullDropZoneDir.length())
        {
            dropzonePathLen = fullDropZoneDir.length();
            dropZone = candidateDropZone;
        }
        break;
    }
```

AttilaVamos replied:
Changed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1698,16 +1697,18 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
-                // Path matched, but machine didn't
-                dropZone = nullptr;
+                    // Keep this candidate drop zone if its directory path is shorter than we already have
+                    if (dropzonePathLen > fullDropZoneDir.length())
```
Should this be >= ?

AttilaVamos replied:
Same as above. Why do you think it would be useful?

jakesmith replied:
It's fine as it is (see prev. comment)

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1827,32 +1828,32 @@ CConstDropZoneServerInfoIterator::CConstDropZoneServerInfoIterator(const IConstD
     bool hasFirst = serverListIt->first();
     if (hasFirst)
-        curr = new CConstDropZoneServerInfo(constEnv, &serverListIt->query());
+        curr.setown(new CConstDropZoneServerInfo(constEnv, &serverListIt->query()));
```
as below in next(), this should clear curr if !hasFirst

AttilaVamos replied:
Added the clear() into the else branch but if there is no first then 'curr' can have any value which should clear?

jakesmith replied:
else curr.clear() is good.
Meaning isValid()  will return false if asked.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -100,7 +100,7 @@ class CConstDropZoneIteratorByComputer : public CSimpleInterfaceOf<IConstDropZon
 {
 public:
     CConstDropZoneIteratorByComputer(const char * computer);
-    ~CConstDropZoneIteratorByComputer() { if (curr != nullptr) delete curr; };
```
delete wasn't correct, but as far as I can see, there needs to be link count handling of curr. It will be leaking an object per first() and next() at the moment.

AttilaVamos replied:
Answered bellow at the ::first().

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1668,22 +1666,23 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
+                           dropZone = candidateDropZone;
+                       }
+                    }
 
```
formatting: extra blank line

AttilaVamos replied:
Removed.

jakesmith commented:
@AttilaVamos - mostly trivial comments, but a couple of issues that do need addressing before merge.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: db3cabc6356b963b1b2d7038f03014b48e54bcc7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.193 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: db3cabc6356b963b1b2d7038f03014b48e54bcc7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.19 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |718 |714 |4 |
|test (thor) |614 |610 |4 |
|test (roxie) |743 |739 |4 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented on common/environment/environment.cpp:
```diff
@@ -1698,16 +1697,18 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
-
-                // Path matched, but machine didn't
-                dropZone = nullptr;
+                    // Keep this candidate drop zone if its directory path is shorter than we already have
```
Changed.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d81f126de2b26fa7063c6ac27f21c67485a6c77a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.575 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith back to you for a final check.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d81f126de2b26fa7063c6ac27f21c67485a6c77a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.043 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |719 |719 |0 |
|test (thor) |614 |614 |0 |
|test (roxie) |744 |744 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1580,53 +1625,95 @@ void CLocalEnvironment::clearCache()
+            else
+            {
+                // It has not '@server' attrib. Check the ServerList
+                bool isServerFound = false;
```
Now unused / should be removed

AttilaVamos replied:
Removed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -59,6 +59,25 @@ class CConstMachineInfoIterator : public  CSimpleInterfaceOf<IConstMachineInfoIt
+    Owned<IConstDropZoneServerInfo> curr;
+    Owned<CLocalEnvironment> constEnv;
+    Owned<IPropertyTreeIterator> serverListIt = nullptr;
+    unsigned index = 1;
```
I think unused / should be removed.

AttilaVamos replied:
Removed.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -59,6 +59,25 @@ class CConstMachineInfoIterator : public  CSimpleInterfaceOf<IConstMachineInfoIt
+protected:
+    Owned<IConstDropZoneServerInfo> curr;
+    Owned<CLocalEnvironment> constEnv;
+    Owned<IPropertyTreeIterator> serverListIt = nullptr;
```
must have missed this one before - inappropriate initializer.

AttilaVamos replied:
Changed.

jakesmith reviewed (commented):
spotted a few more minors.

jakesmith commented:
@AttilaVamos - spotted a few minors.
Could be merged now, but should be squashed.

AttilaVamos commented:
@jakesmith last minute small things changed, commits are squashed.

jakesmith commented:
@richardkchapman - good to merge.

jakesmith commented:
@richardkchapman - please don't merge yet.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 1957cc198d78cbef6b77d6ff7fde61f7b8601495
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 95.008 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |719 |718 |1 |
|test (thor) |614 |614 |0 |
|test (roxie) |744 |744 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
The Hthor error is a know issue related to --pq
```
436. Test: persist_refresh.ecl ( version: persistRefresh=false )
    '--- persist_refresh.xml
    +++ persist_refresh.xml
```

jakesmith commented:
Re. @wangkx comments in JIRA (how to get a DZ with IP only)..

That highlights that the getDropZoneByComputer* and getDropZoneIteratorByComputer methods are legacy and shouldn't be used - will not work with new style dropzones.

The new methods should be used which are backward compatible, i.e. the generic iterator, or getDropZoneByAddressPath.

The only use case of getDropZoneByComputer now, is in Esp service code during a despray, when a relative path has been provided.
@wangkx - could use the getDropZoneIterator() to find a IP match, though the code I think was problematic before because it assumed only 1 DZ would ever be defined that matched and there could have been more.

jakesmith commented:
It may be sensible to add a new iterator method or add an optional ip/netaddress param to the existing getDropZoneIterator() to allow just DropZone's that reside on the given IP to be returned.
But I'd speak to @wangkx about his requirements and how that's going to work 1st.

AttilaVamos commented:
@jakesmith Legacy DropZone @computer handling in CConstDropZoneServerInfoIterator added. Please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 35857ab3901cfc24b793fce286b0f28446e2feb9
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.758 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |720 |719 |1 |
|test (thor) |615 |615 |0 |
|test (roxie) |745 |745 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d42cc967ebd631ecb861fe9ba231ff41d38aef25
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.244 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |720 |720 |0 |
|test (thor) |615 |615 |0 |
|test (roxie) |745 |745 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1821,9 +1791,45 @@ CConstDropZoneServerInfoIterator::CConstDropZoneServerInfoIterator(const IConstD
+            // Create a single ServerList record related to @computer
+            //<ServerList name="ServerList" server="<IP_of_@computer>"/>
+            Owned<IPropertyTree> newRecord = createPTree();
+            newRecord->setProp("@name", "ServerList");
```
minor: the computer name would be more appropriate here I think (dropZoneMachineName) (and similarly in comment above)

AttilaVamos replied:
I kept "ServerList" as name because actually I don't know who, where and how to plan to use the content of the name property, I tried to make backward compatibility transparent and maybe it can cause trouble in the future if the name is not "ServerListN".
However, if you wish to change I can do that, but in this case I think the related documentation should change as well to emphasize the difference.

jakesmith replied:
You could check with @rpastrana, I think they're completely arbitrary.
Also, the default is "ServerListEntry" not "ServerList" afaics.

jakesmith replied:
@AttilaVamos - I've checked with @rpastrana , it doesn't matter what it is,  configmgr / xslt generation needed it to be a unique something, but here for your internally generated version it can be anything.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -75,6 +75,7 @@ class CConstDropZoneServerInfoIterator : public CSimpleInterfaceOf<IConstDropZon
     Owned<CLocalEnvironment> constEnv;
     Owned<IPropertyTreeIterator> serverListIt;
     unsigned maxIndex = 0;
+    Owned<IPropertyTree> legacyServerList;
```
not needed as member, getElements will link parent it's iterating.
So this could be moved into if (0 != dropZoneMachineName.length()) block in CConstDropZoneServerInfoIterator constructor.

AttilaVamos replied:
Moved to the implementation

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1821,9 +1791,45 @@ CConstDropZoneServerInfoIterator::CConstDropZoneServerInfoIterator(const IConstD
+    if (0 != dropZoneMachineName.length())
+    {
+        // Create a ServerList for legacy element.
+        legacyServerList.setown(createPTree());
```
See previous comment, better to declare legacyServerList here, getElements will link it, so retrained/valid because serverListIt is kept.

AttilaVamos replied:
Changed.

jakesmith commented:
@AttilaVamos - looks good, a couple of minor comments.

AttilaVamos commented:
@jakesmith updated and one comment answered.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: fceee9f0091e3903113fd7ff90fd5acafcb46288
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.919 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |720 |719 |1 |
|test (thor) |615 |615 |0 |
|test (roxie) |745 |745 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
Hthor error is a well known persist_refresh.ecl parallel query problem.

jakesmith commented:
@richardkchapman - ready to merge.

richardkchapman commented:
@AttilaVamos Can you squash

AttilaVamos committed:
HPCC-16946 Add server list handling to CLocalEnvironment

Reorganise checking process

Add Drop Zone server list checking if the target machine doesn't match
to local one.

Create IConstDropZoneServerInfoIterator interface to easy handle drop
zone server list.

Add legacy DropZone @computer handling

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@richardkchapman done.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 18b3c17152cf522f0b31fc3ae3190ed0fb0db775
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.829 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |720 |720 |0 |
|test (thor) |615 |615 |0 |
|test (roxie) |745 |745 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
